Class-based UDFs

sk1p opened the pull request:
Not finished, for discussion of the new interface. Refs #286

uellue commented on src/libertem/io/dataset/base.py:
```diff
@@ -289,7 +289,11 @@ def get_tiles(self, crop_to=None, full_frames=False, mmap=False, dest_dtype="flo
 
         roi : np.ndarray
-            1d mask that matches the partition shape to limit the region to work on
+            1d mask that matches the dataset navigation shape to limit the region to work on.
```
Reminder for @sk1p: Update comment

uellue commented on src/libertem/udf/base.py:
```diff
@@ -1,6 +1,275 @@
+
+
+class UDF(UDFBase):
+    def __init__(self, **kwargs):
```
Draft: This method is called again on the cluster to create copies. For that reason it should not have side effects beyond `self` and only depend on its parameters.

Discussion with @sk1p: Overloading this method with sth fancy may have unwanted side effects.

Documentation: Include a clear example to condition parameters.

sk1p commented on src/libertem/udf/FEM.py:
```diff
@@ -1,44 +1,42 @@
+        self.results.intensity[:] = np.std(frame[self.task_data.mask == 1])
+
+
+def run_fem(ctx, dataset, center, rad_in, rad_out, roi=None):
```
Hmm, I would say, we should defer this change until we have the integration with the web API, which should also handle multi-stage UDFs for scripting. That means UDFs developed for 0.2 will need a little work to be available in the web API, but I think that's fine.

Maybe this becomes more clear if we port over the blobfinder UDF to the new interface.

sk1p commented:
I addressed most of the comments, only the async/web API stuff remains. Also, the `full_frames=False` tiled processing is not finished, see #310 for a remark about possible interface changes there.

Do you want to port over the blobfinder UDFs, or should I do this?

sk1p commented on src/libertem/udf/base.py:
```diff
@@ -9,54 +278,64 @@ def check_cast(fromvar, tovar):
+        tasks = self._make_udf_tasks(dataset, roi)
+        cancel_id = str(uuid.uuid4())
+
+        for part_results, partition in executor.run_tasks(tasks, cancel_id):
```
There is now a first prototype of an async version of this function available, including integration with web API

sk1p commented:
`udf_v2_wip` is now merged!
